client/main.go

```diff
@@ -229,118 +254,133 @@ func loadPicture(path string) (pixel.Picture, error) {
 
 func requestMove(direction shared.Direction, conn net.Conn) error {
 	msg := &shared.Message{
-		MoveRequest: &shared.MoveRequest{
+		Request: &shared.Request{MoveRequest: &shared.MoveRequest{
 			Direction: direction,
 			Created:   time.Now(),
 		},
-	}
+		}}
 	return shared.SendMessage(msg, conn)
 }
 
 func requestSpeak(txt string, conn net.Conn) error {
 	msg := &shared.Message{
-		SpeakRequest: &shared.SpeakRequest{
+		Request: &shared.Request{SpeakRequest: &shared.SpeakRequest{
 			Text: txt,
-		},
+		}},
 	}
 	return shared.SendMessage(msg, conn)
 }
 
-func handleConnection(conn net.Conn) {
+func (g *GameWorld) handleConnection(conn net.Conn) {
 	loop := func() error {
 		msg, err := shared.GetMessage(conn)
 		if err != nil {
 			return err
 		}
-		switch {
-		case msg.PlayerMoved != nil:
-			handlePlayerMoved(msg.PlayerMoved)
-		case msg.PlayerSpoke != nil:
-			handlePlayerSpoke(msg.PlayerSpoke)
-		case msg.WorldState != nil:
-			handleWorldState(msg.WorldState)
-		case msg.PlayerDisconnected != nil:
-			handlePlayerDisconnected(msg.PlayerDisconnected)
+		log.Println("RECV", msg)
+		if msg.Update != nil {
+			g.ApplyUpdate(msg.Update)
 		}
 		return nil
 	}
 	for {
 		if err := loop(); err != nil {
-			errc <- err
+			g.errc <- err
 			continue
 		}
+
 	}
 }
 
-func handlePlayerMoved(moved *shared.PlayerMoved) {
-	setPlayerPosition(moved.ID, moved.NewPosition)
-	reapplySimulations(moved.RequestTime)
+func (g *GameWorld) ApplyUpdate(update *shared.Update) {
```

Comment:
I scrapped the `switch` so that updates can have more than one update type in the future